pkg/log/flags.go

```diff
@@ -54,283 +47,319 @@ var LevelNames = map[Level]string{
 		}
-		o.parseFlags(optionSource, sources, state)
+	}
 
-	case optionDebug:
-		sources, state, err := o.splitFlagsAndState(value)
-		if err != nil {
-			return err
-		}
-		o.parseFlags(optionDebug, sources, state)
+	return nil
+}
 
-	case optionPrefix:
-		switch value {
-		case "off", "false":
-			o.prefix = 0
-		default:
-			o.prefix = 1
+func (m *stateMap) String() string {
+	if *m == nil {
+		return "all"
+	}
+	if len(*m) == 0 {
+		return "none"
+	}
+
+	tVal, tSep := "", ""
+	fVal, fSep := "", ""
+
+	for name, state := range *m {
+		if name == "*" {
+			name = "all"
+		}
+		if state {
+			tVal += tSep + name
+			tSep = ","
+		} else {
+			fVal += fSep + name
+			fSep = ","
 		}
+	}
 
-	case optionList:
-		fmt.Printf("The available logger backends are: %s\n", registeredBackendNames())
-		os.Exit(0)
+	if tVal != "" {
+		tVal = "on:" + tVal
+	}
+	if fVal != "" {
+		fVal = "off:" + fVal
+	}
+
+	switch {
+	case fVal == "":
+		return tVal
+	case tVal == "":
+		return fVal
+	case tVal != "" && fVal != "":
+		return tVal + "," + fVal
+	}
+	return ""
+}
 
-	default:
-		return fmt.Errorf("can't set unknown logger option '%s' to '%s'", name, value)
+func (m *stateMap) isEnabled(name string) bool {
+	if m == nil || *m == nil {
+		return false
 	}
 
-	return nil
+	if state, ok := (*m)[name]; ok {
+		return state
+	}
+	if state, ok := (*m)["*"]; ok {
+		return state
+	}
+
+	return false
 }
 
-func (o *options) Get(name string) string {
-	switch name {
-	case optionLogger:
-		return o.logger
+func (o *options) MarshalJSON() ([]byte, error) {
+	cfg := map[string]string{
+		"Level":  o.level.String(),
+		"Logger": string(o.logger),
+	}
+	if o.enable != nil {
+		cfg["Enable"] = o.enable.String()
+	}
+	if o.debug != nil {
+		cfg["Debug"] = o.debug.String()
+	}
 
-	case optionLevel:
-		return o.level.String()
+	return json.Marshal(cfg)
+}
 
-	case optionSource:
-		switch {
-		case o.sources == nil:
-			return "all"
-		case len(o.sources) == 0:
-			return "none"
-		default:
-			value := ""
-			sep := ""
-			for source, state := range o.sources {
-				value += sep + source + ":" + strconv.FormatBool(state)
-			}
-			return value
-		}
+func (o *options) UnmarshalJSON(raw []byte) error {
+	cfg := map[string]string{}
 
-	case optionDebug:
-		switch {
-		case o.debugs == nil:
-			return "all"
-		case len(o.debugs) == 0:
-			return "none"
-		default:
-			value := ""
-			sep := ""
-			for source, state := range o.debugs {
-				value += sep + source + ":" + strconv.FormatBool(state)
-			}
-			return value
-		}
+	if err := json.Unmarshal(raw, &cfg); err != nil {
+		return loggerError("failed to unmarshal logger configuration: %v", err)
+	}
 
-	case optionPrefix:
-		switch o.prefix {
-		case 0:
-			return "off"
-		case 1:
-			return "on"
+	*o = *defaults
+
+	for key, value := range cfg {
+		switch key {
+		case "Level":
+			if err := o.level.Set(value); err != nil {
+				return err
+			}
+		case "Logger":
+			o.logger = backendName(value)
+			activateBackend(value)
```

Comment:
It's a bit  peculiar that the options are acivated applied in the json unmarshal phase when config parsing is still underway. But I guess logger is a special case and this is, if not inevitable, easier this way(?)

Reply:
In principle we could delay reconfiguring the logger until the notification callback comes in. But I agree that logging is a bit special and that's the reason we try to make a successful switch as early as possible.

Reply:
And I guess if this turns out to be confusing/disturbing in case of rollbacks after failure, we can always switch to reconfiguring from the callback.